addons/mail/models/mail_channel.py

```diff
@@ -798,13 +799,12 @@ def channel_search_to_join(self, name=None, domain=None):
     @api.multi
     def channel_join_and_get_info(self):
         self.ensure_one()
-        if self.channel_type == 'channel' and not self.email_send:
+        added = self.action_follow()
+        if added and self.channel_type == 'channel' and not self.email_send:
             notification = _('<div class="o_mail_notification">joined <a href="#" class="o_channel_redirect" data-oe-id="%s">#%s</a></div>') % (self.id, self.name,)
             self.message_post(body=notification, message_type="notification", subtype="mail.mt_comment")
-        self.action_follow()
-
-        if self.moderation_guidelines:
-            self._send_guidelines(self.env.user.partner_id)
+            if self.moderation_guidelines:
+                self._send_guidelines(self.env.user.partner_id)
 
         channel_info = self.channel_info('join')[0]
         self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', self.env.user.partner_id.id), channel_info)
```

Comment:
We are sending a notification on bus AND returning the value to current user.
-> if called from a button, message on bus is useful but not return 
-> if called from _joinAndAddChannel, return should be enough

I think we may add a args on channel_join_and_get_info to avoid on useless bus notification when called from there, but this isn't the scope of this fix.